### PR TITLE
Add criticality prediction

### PR DESCRIPTION
**Pull request by Stern5497**
Add a new task criticality prediction.
- it uses it's own script and is no default task like sltc
- another helper class was added to add util functions for criticality only
- run_ubelix.sh file was added to run on ubelix cluster

Add new task citation prediction

### CODE REVIEW

File: experiments/template_SLTC.py [hunk 1]
```diff
@@ -220,6 +235,7 @@ def main():
         trainer.log_metrics("train", metrics)
         trainer.save_metrics("train", metrics)
         trainer.save_state()
+        wandb.log(metrics)
```
Comment (kapllan) [inline experiments/template_SLTC.py]: Is this step really necessary? Doesn't the logging happen with trainer.log_metrics?

File: main.py [hunk 2]
```diff
@@ -283,13 +283,17 @@ def run_experiment(
             if accumulation_steps is None:
                 accumulation_steps = 1
 
+    # TODO this causes errors because model_language_mapping is not in mate_info
```
Comment (kapllan) [inline main.py]: @Stern5497 are you sure this is the case? This field added, see: https://github.com/JoelNiklaus/LEXTREME/blob/main/utils/utilities.py#L41

**Review by kapllan — APPROVED** [review]
I think the changes are ok. It works on my machine. They can be merged.

File: experiments/helper.py [hunk 3]
```diff
@@ -84,6 +84,16 @@ def make_efficient_split(data_args, split_name, ner_tasks):
     return dataset
 
 
+def filter_by_length(row):
+    """
+    Removes examples that are too short
+    :param example:    the example to check
+    :return:
+    """
+    if 10 < len(str(row["input"])):
```
Comment (JoelNiklaus) [inline experiments/helper.py]: This might lead to different results on LEXTREME if we rerun the models. In some datasets we didn't filter out these very small ones before.

File: experiments/template_SLTC.py [hunk 4]
```diff
@@ -64,6 +66,12 @@ def main():
     if data_args.disable_caching:
         disable_caching()
 
+    if data_args.finetuning_task in ["swiss_criticality_prediction_bge_facts",
+                                     "swiss_criticality_prediction_bge_considerations",
+                                     "swiss_criticality_prediction_citation_facts",
+                                     "swiss_criticality_prediction_citation_considerations"]:
+        os.environ["WANDB_API_KEY"] = "6f9fce3d2b3e41f8880b3e0b094e16ec9d030315"  # wandb rstern token
```
Comment (JoelNiklaus) [inline experiments/template_SLTC.py]: there should never be any secrets on GitHub in the open!
This should be put in an .env file for example and loaded from there

File: experiments/template_SLTC.py [hunk 5]
```diff
@@ -122,6 +130,11 @@ def main():
     set_seed(training_args.seed)
 
     train_dataset, eval_dataset, predict_dataset = get_data(training_args, data_args)
+    train_dataset = train_dataset.filter(lambda example: len(example['input']) > 0)
+    eval_dataset = eval_dataset.filter(lambda example: len(example['input']) > 0)
+    predict_dataset = predict_dataset.filter(lambda example: len(example['input']) > 0)
+
+    print(train_dataset['label'])
```
Comment (JoelNiklaus) [inline experiments/template_SLTC.py]: is this a debug output?

File: experiments/template_SLTC.py [hunk 6]
```diff
@@ -134,6 +147,11 @@ def main():
         label2id[l] = n
         id2label[n] = l
 
+    if 'citation' in data_args.finetuning_task:
+        train_dataset = train_dataset.filter(lambda row: str(id2label[row['label']]).startswith("critical"))
+        eval_dataset = eval_dataset.filter(lambda row: id2label[row['label']].startswith("critical"))
+        predict_dataset = predict_dataset.filter(lambda row: id2label[row['label']].startswith("critical"))
+
```
Comment (JoelNiklaus) [inline experiments/template_SLTC.py]: What does this do?

File: run_ubelix_job.sh [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+#SBATCH --mail-type=end,fail
+#SBATCH --nodes=1
+#SBATCH --ntasks-per-node=1
+#SBATCH --mem=64GB
+#SBATCH --cpus-per-task=8
+#SBATCH --gres=gpu:rtx3090:1
+#SBATCH --time=24:00:00
+#SBATCH --qos=job_gpu_preempt
+#SBATCH --partition=gpu
+#SBATCH --array=2-4
```
Comment (JoelNiklaus) [inline run_ubelix_job.sh]: we run seeds 1-3

File: run_ubelix_job.sh
Comment (JoelNiklaus) [inline run_ubelix_job.sh:1]: This file doesn't need to be committed. We can think about committing a template though

File: utils/meta_infos.json [hunk 8]
```diff
@@ -56,7 +68,12 @@
     "legalnero": "LNR",
     "lener_br": "LNB",
     "mapa": "MAP",
-    "turkish_constitutional_court_decisions_judgment": "TCD"
+    "turkish_constitutional_court_decisions_judgment": "TCD",
+    "criticality_prediction": "CP",
```
Comment (JoelNiklaus) [inline utils/meta_infos.json]: should this be swiss_criticality_prediction: SCP?

Comment (JoelNiklaus) [reply utils/meta_infos.json]: And the sub datasets should be removed here, right?

File: utils/meta_infos.json [hunk 9]
```diff
@@ -277,6 +302,22 @@
     ],
     "turkish_constitutional_court_decisions_judgment": [
       "tr"
+    ],
+    "swiss_criticality_prediction_bge_facts": [
+      "all"
+    ],
+    "swiss_criticality_prediction_bge_considerations": [
+      "all"
```
Comment (JoelNiklaus) [inline utils/meta_infos.json]: should be replaced by de,fr,it right?

File: utils/meta_infos.json [hunk 10]
```diff
@@ -298,7 +339,11 @@
     "lener_br": "lener_br",
     "mapa_coarse": "mapa",
     "mapa_fine": "mapa",
-    "turkish_constitutional_court_decisions_judgment": "turkish_constitutional_court_decisions_judgment"
+    "turkish_constitutional_court_decisions_judgment": "turkish_constitutional_court_decisions_judgment",
+    "swiss_criticality_prediction_bge_facts": "swiss_criticality_prediction_bge_facts",
+    "swiss_criticality_prediction_bge_considerations": "swiss_criticality_prediction_bge_considerations",
+    "swiss_criticality_prediction_citation_facts": "swiss_criticality_prediction_citation_facts",
+    "swiss_criticality_prediction_citation_considerations": "swiss_criticality_prediction_citation_considerations"
```
Comment (JoelNiklaus) [inline utils/meta_infos.json]: here they should all map to swiss_criticality_prediction

File: utils/meta_infos.json [hunk 11]
```diff
@@ -343,6 +388,18 @@
...
+     "swiss_criticality_prediction_bge_facts"
+  ],
+    "swiss_criticality_prediction_bge_considerations": [
+      "swiss_criticality_prediction_bge_considerations"
+    ],
+    "swiss_criticality_prediction_citation_facts": [
+      "swiss_criticality_prediction_citation_facts"
+    ],
+    "swiss_criticality_prediction_citation_considerations": [
+      "swiss_criticality_prediction_citation_considerations"
```
Comment (JoelNiklaus) [inline utils/meta_infos.json]: adapt this accordingly too

File: utils/meta_infos.json [hunk 12]
```diff
@@ -523,6 +580,50 @@
...
+    "swiss_criticality_prediction_citation_considerations": {
+      "DataTrainingArguments": {
+        "max_seq_length": 512,
+        "max_segments": 16,
+        "max_seg_length": 128,
+        "add_oversampling": true
+      },
+      "ModelArguments": {
+        "hierarchical": true
+      }
```
Comment (JoelNiklaus) [inline utils/meta_infos.json:611]: Is there a way to make all of this less "boilerplaty"?

**Review by JoelNiklaus — COMMENTED** [review]
Thanks for the PR. Looks like its is on a good track already. Please make sure, it integrates nicely with how the repo is structured already. I left some comments that might help here :)

File: utils/meta_infos.json (hunk 12 above)
Comment (kapllan) [reply utils/meta_infos.json:611]: @JoelNiklaus before that, as far as I remember, we saved these information in main.py. I thought this was a nice solution to define these arguments for each dataset.

Comment (JoelNiklaus) [reply utils/meta_infos.json:611]: Yes, I definitely agree. It is a good solution to pull all of this information out into a config file.
Now we have the problem of lots of config duplication: The config of swiss_criticality_prediction_bge_considerations is identical to swiss_criticality_prediction_citation_facts. We could for example copy default dicts and only change the difference (e.g. here: https://huggingface.co/datasets/joelito/lextreme/blob/main/lextreme.py#L132). 
Or we could use classes and inheritance.

We want to keep the max_segment_length constant over all the scenarios. So this does not need to go here for every dataset.
We can infer hierarchical based on the sequence length (true if larger than 512). So we have potential to make this cleaner :)

File: utils/meta_infos.json (hunk 8 above)
Comment (Stern5497) [reply utils/meta_infos.json]: I think line 72 is unnecessary.

File: utils/meta_infos.json (hunk 10 above)
Comment (Stern5497) [reply utils/meta_infos.json]: I thought this is where I define the dataset config I want to use. If all map to swiss_criticality_prediction how can i use a different dataset config?

File: run_ubelix_job.sh
Comment (Stern5497) [reply run_ubelix_job.sh:1]: File is removed

File: run_ubelix_job.sh (hunk 7 above)
Comment (Stern5497) [reply run_ubelix_job.sh]: Done

File: experiments/template_SLTC.py (hunk 6 above)
Comment (Stern5497) [reply experiments/template_SLTC.py]: Here I filter only critical cases, it makes sure only using cases with labels critical-1 critical-2 critical-3 or critical-4. I guess it is reduntant because we implemented it in the lextreme loading script.

File: experiments/template_SLTC.py (hunk 5 above)
Comment (Stern5497) [reply experiments/template_SLTC.py]: Not really neccesary, but I thougt it's nice to have it in the output file. I can remove it.

File: experiments/template_SLTC.py (hunk 4 above)
Comment (Stern5497) [reply experiments/template_SLTC.py]: ok

File: utils/meta_infos.json (hunk 10 above)
Comment (Stern5497) [reply utils/meta_infos.json]: I got it wrong and corrected it. We have now Swiss_Criticality_Prediction with 4 different tasks.

File: experiments/template_SLTC.py (hunk 5 above)
Comment (JoelNiklaus) [reply experiments/template_SLTC.py]: No it's fine. But please use the logger instead of print and add some more information

File: experiments/template_SLTC.py (hunk 6 above)
Comment (JoelNiklaus) [reply experiments/template_SLTC.py]: Yeah I think it's not necessary anymore here too. I think removing it here is better to keep complexity low.

File: experiments/helper.py (hunk 3 above)
Comment (JoelNiklaus) [reply experiments/helper.py]: But we can still keep filtering completely empty examples...

**Comment by JoelNiklaus** [comment]
Thanks for the updates. Looks much better now, thanks.

File: experiments/helper.py (hunk 3 above)
Comment (Stern5497) [reply experiments/helper.py]: We do filter completely empty examples directly in the template_SLTK.py (see line 127-129). This function was not used.

File: experiments/template_SLTC.py (hunk 5 above)
Comment (Stern5497) [reply experiments/template_SLTC.py]: ok :)

File: experiments/template_SLTC.py (hunk 6 above)
Comment (Stern5497) [reply experiments/template_SLTC.py]: ok

File: experiments/helper.py (hunk 3 above)
Comment (JoelNiklaus) [reply experiments/helper.py]: Ooh ok, great.

File: experiments/template_SLTC.py (hunk 5 above)
Comment (JoelNiklaus) [reply experiments/template_SLTC.py]: thanks :)
